Test/Utils/test_struct.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2021 The Flax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for flax.struct."""
+
+
+from .. import common
+
+pytestmark = common.skipif_mpi
+
+from typing import Any
+import pytest
+
+import dataclasses
+
+from netket.utils import struct
+
+import jax
+
+
+@struct.dataclass
+class Point0:
+    x: float
+    y: float
+    meta: Any = struct.field(pytree_node=False)
+
+    def __pre_init__(self, *args, **kwargs):
+        if "z" in kwargs:
+            kwargs["x"] = kwargs.pop("z") * 10
+
+        return args, kwargs
```

Comment:
What happens if "z" is not passed as a positional argument here, by the way?

Reply:
If it’s not a positional argument you cannot pass ‘z’ explicitly, but you’ll be passing x and y respectively?

you can override the positional arguments if you want and say that the first argument is z.

You just need in the end to pass what the data class wants .

I do this for sampler/base and metropolis.

Reply:
Hm, okay in this case, since `Point` has no field `z` I guess it has to be passed as kwarg in any case.

But what happens if I have
```python
@struct.dataclass
class Point:
    x: int
    y: int
```
and want to preprocess `y` in some way. Since both `Point(x, y)` and `Point(x=x, y=y)` are valid ways to construct the dataclass, do I, in `__pre_init__`, have to account for the fact that the "y" I want to preprocess can be in both `*args` and `**kwargs`? Or do I have to use `__pre_init__(x, y, *args, **kwargs)` to catch `y` correctly?

Reply:
In the snippet above, data class will generate the following `__init__` method
```python
class Point:
  def __init__(self, x, y):
    self.__setattr__(self, 'x', x)
    self.__setattr__(self, 'y', y)
```

What my Logic does is that it replaces this by something like 
```python
class Point:
  def __init__(self, *args, **kwargs):
     if hasattr(self, __pre_init__):
        args, kwargs = self.__pre_init__(*args, **kwargs)
    self. __pre_init__(*args, **kwargs)

  def __init_dataclass__(self, x, y):
    self.__setattr__(self, 'x', x)
    self.__setattr__(self, 'y', y)
```

So `__pre_init__` has to spit out something that can be parsed by the data class constructor: it can be anything among `(x,y), {}`, `(,),{'y':y}`, `(,),{'x':x, 'y':y}`.

What you decide to accept in the input is up to you.
In your example, you might do 

```python
__pre_init__(self, x,y):
   x = x*10
   return (x,y)
```

Reply:
Okay, sounds good!